node/core/candidate-validation/src/lib.rs

```diff
@@ -459,7 +472,15 @@ async fn validate_candidate_exhaustive(
 			Ok(ValidationResult::Invalid(InvalidCandidate::ExecutionError(
 				"ambigious worker death".to_string(),
 			))),
-
+		Err(ValidationError::ArtifactNotFound) => {
+			// The code was supplied on the second attempt, this
+			// error should be unreachable.
+			let err = ValidationFailed(
+				"Validation host failed to find artifact even though it was supplied".to_string(),
+			);
+			tracing::error!(target: LOG_TARGET, error = ?err, "Unexpected error reported by the validation backend");
+			Err(err)
```

Comment:
This rubs me the wrong way. There are too many strings invovled here and `ValidationFailed` should become a proper error enum, but that's out of scope for this PR. I also don't see a point to print the error here, the caller should take care of printing the `Err(err)`.

Reply:
In fact, this could be replaced with `unreachable`, but since we try to avoid panicking the error simply gets logged. Caller (as I understand the component that sent the validation request?) should never receive this error, so we handle it here. Though I totally agree on the part of turning strings into enum. 

The PR is on ice due to dependency on the recent runtime upgrade, so probably I could fix `ValidationFailed` too, but I wonder how would anybody handle variants? Isn't it a kind of internal error that means something went wrong and we should skip the candidate?

Reply:
What my comment boils down to, is that 

```

/// Blanket error for validation failing for internal reasons.
#[derive(Debug, Error)]
#[error("Validation failed with {0:?}")]
pub struct ValidationFailed(pub String);
```
exists which is an error antipattern, and makes the code less legible than it could be. It should be an error enum with explicit variants and annotations using `thiserror`.